### PR TITLE
Add GPU optimization for segresnet

### PR DESCRIPTION
**Pull request by dongyang0122**
Signed-off-by: dongy <dongy@nvidia.com>

### CODE REVIEW

File: auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml [hunk 1]
```diff
@@ -23,7 +23,7 @@ cache_rate: null
 determ: false
 fold: 0
 learning_rate: 2.0e-4
-num_images_per_batch: 1
+batch_size: 1
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: If I remember correctly, the other 3 algorithms use `num_images_per_batch`.

Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: it is named as "batch_size" in "train.py" of segresnet. I did not change anything yet in the training script since I am not sure if naming serves any purpose in segresnet. @myron

Comment (dongyang0122) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: I just unify the naming in "algo.py".

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py [hunk 2]
```diff
@@ -74,19 +90,25 @@ def fill_template_config(self, data_stats_file: Optional[str] = None, output_pat
...
                 input_config = ConfigParser.load_config_file(self.data_list_file)
                 print("Loaded self.data_list_file", self.data_list_file)
             else:
                 print("Unable to load self.data_list_file", self.data_list_file)
 
             config["data_file_base_dir"] = os.path.abspath(input_config.pop("dataroot"))
-            config["data_list_file_path"] = os.path.abspath(input_config.pop("datalist"))
+            config["data_list_file_path"] = os.path.abspath(
+                input_config.pop("datalist")
+            )
```
Comment (myron) [inline auto3dseg/algorithm_templates/segresnet/scripts/algo.py:111]: why breaking down so many single lines into these many lines?

File: auto3dseg/algorithm_templates/segresnet/scripts/algo.py [hunk 3]
```diff
@@ -110,7 +134,10 @@ def fill_template_config(self, data_stats_file: Optional[str] = None, output_pat
             # update config config
             roi_size = [224, 224, 144]  # default roi
             levels = 5  # default number of hierarchical levels
-            image_size = [int(i) for i in data_stats["stats_summary#image_stats#shape#percentile_99_5"]]
+            image_size = [
```
Comment (myron) [inline auto3dseg/algorithm_templates/segresnet/scripts/algo.py:137]: this looks much less compact

File: auto3dseg/algorithm_templates/segresnet/scripts/dummy_runner.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                if np.prod(new_shape) > np.prod(self.max_shape):
+                    self.max_shape = new_shape
+        else:
+            for _k in range(len(data_stat["stats_by_cases"])):
+                image_shape = data_stat["stats_by_cases"][_k]["image_stats"]["shape"]
+                image_shape = np.squeeze(image_shape)
+
+                if np.prod(image_shape) > np.prod(self.max_shape):
+                    self.max_shape = image_shape
+        print("max_shape", self.max_shape)
```
Comment (myron) [inline auto3dseg/algorithm_templates/segresnet/scripts/dummy_runner.py:90]: you can compute max_shape first, and then account for pix_dim later (if resample)

also this can be inaccurate since we pre-crop images on CPU first and the shape changes. We crop foreground always, and crop a special region for a recent tutorial.  You could use "image_size" from a config (which is an approximate estimate)

File: auto3dseg/algorithm_templates/segresnet/scripts/dummy_runner.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                )
+                labels = torch.rand(
+                    (
+                        num_images_per_batch,
+                        self.label_channels,
+                        self.roi_size[0],
+                        self.roi_size[1],
+                        self.roi_size[2],
+                    )
+                )
```
Comment (myron) [inline auto3dseg/algorithm_templates/segresnet/scripts/dummy_runner.py:140]: since the dummy label is not an index map (0,1,2,3,..) it won't be correctly one_hot encoded during loss calculation

**Comment by myron** [comment]
for Segresnet algo only

1. Overall I think it would be much easier to re-use existing segmenter.py class to run train_epoch() or val_epoch() on a random data, then to write a dummy code (repeating most of the code,  if we ever change something in the main class, we'll always have to worry about updating the dummy class the same way, e.g.self. sliding_inferer)

2. unfortunately, given a short amount of time, It won't be possible for me to properly test it for segresnet...   so I recommend, to either wait  for a longer time... Or exclude changes  to segresnet algo from this PR, and make a separate PR for that  (which I can review in details)

thank you guys

File: auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml (hunk 1 above)
Comment (myron) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: yes, I've checked and I'm actually using "batch_size" everywhere, so this proposed change is correct.. but ideally we want to bring this name to be the same to all algos (either one can be ok)

**Comment by mingxin-zheng** [comment]
Correct me if I am wrong @wyli ：
Since 1.1 is in code-freeze, this PR won't go into the formal release. For users that want to use these features (segresnet gpu optimization, change num_workers for algorithms other than segresnet, ...), they have to set the environment variable `ALGO_HASH` in their systems.

File: auto3dseg/algorithm_templates/dints/scripts/search.py [hunk 6]
```diff
@@ -150,56 +159,42 @@ def run(config_file: Optional[Union[str, Sequence[str]]] = None, **override):
...
+        transform=val_transforms,
+        cache_rate=validate_cache_rate,
+        num_workers=2,
+        progress=False,
+    )
 
     train_loader_a = ThreadDataLoader(
-        train_ds_a, num_workers=6, batch_size=num_images_per_batch, shuffle=True
+        train_ds_a,
+        num_workers=parser.get_parsed_content("searching#num_workers"),
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/dints/scripts/search.py:189]: Can we use `num_workers` for both searching and training in DiNTS?

Otherwise the user has to override the training_params with something like
```
train_params = {"num_workers": 4, "searching#num_workers": 4, "training#num_workers": 4}
```
in `AutoRunner`

File: auto3dseg/algorithm_templates/dints/configs/hyper_parameters_search.yaml [hunk 7]
```diff
@@ -19,6 +23,7 @@ searching:
   num_warmup_epochs: 500
   num_patches_per_image: 1
   num_sw_batch_size: 2
+  num_workers: 6
```
Comment (mingxin-zheng) [inline auto3dseg/algorithm_templates/dints/configs/hyper_parameters_search.yaml:26]: I am wondering why we use 8 workers in training but 6 workers in searching?

File: auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml (hunk 1 above)
Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: On another note, maybe we should make a note about the unified names, such as `num_epochs` and `batch_size`/`num_images_per_batch` in this or the `monai.apps.auto3dseg`

Comment (mingxin-zheng) [reply auto3dseg/algorithm_templates/segresnet/configs/hyper_parameters.yaml:26]: I would propose let's fix it to `num_images_per_batch` so that the tests/tutorials work as before:

https://github.com/Project-MONAI/research-contributions/blob/c812e5f3b7a56c998035514846aa3292394c4097/auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py#L712

**Comment by mingxin-zheng** [comment]
Hi @dongyang0122 , can we also expose the "num_workers" when `CacheDataset` is used? Looks like user experience can be much improved by reducing it on machines with low-end CPUs. Also, we need to distinguish these `num_workers` in CacheDataset and DataLoader in the configs.
